spec/formatting.md

```diff
@@ -19,6 +19,199 @@ the local variable takes precedence.
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
 
+## Pattern Selection
+
+When formatting a message with one or more _selectors_,
+the _pattern_ of one of the _variants_ must be selected for formatting.
+
+When a message has a single _selector_,
+an implementation-defined method compares each key to the _selector_
+and determines which of the keys match, and in what order of preference.
+A catch-all key will always match, but is always the least preferred choice.
+During selection, the _variant_ with the best-matching key is selected.
+
+In a message with more than one _selector_,
+each _variant_ also has a corresponding number of keys.
+These correspond to _selectors_ by position.
+The same implementation-defined method as above is used to compare
+the corresponding key of each _variant_ to its _selector_,
+to determine which of the keys match, and in what order of preference.
+In order to select a single _variant_,
+the full list of _variants_ will need to be filtered and sorted.
+First, each _variant_ with a key that does not match its _selector_ is left out.
+Then, the remaining _variants_ are sorted lexicographically by key preference,
+with earlier _selectors_ having higher priority than later ones.
+Finally, the highest-sorted _variant_ is selected.
+
+This selecion method is defined in more detail below.
+An implementation MAY use any pattern selection method,
+as long as its observable behaviour matches the results of the method defined here.
+
+### Resolve Selectors
+
+First, resolve the values of each _selector_:
```

Comment:
I don't understand this? What are the "values" of a _selector_?

I think a _selector_ has two inputs: an _expression_ (either a variable or a literal) and a list of keys. Resolving the "value" of the _expression_ seems to be early-binding in a way that isn't required to implement the spec. Or is "value" something else?

Reply:
Happy to play with the semantics here, but what I intend the "resolved value" here to mean is probably pretty close to your understanding of a _selector_ as a function. For example, let's say that we start from the following syntax representation of a selector:
```
{$count :plural minimumFractionDigits=2}
```
This may be parsed e.g. into the following data model representation, which we can still claim to represent the _selector_:
```js
{
  type: 'expression',
  name: 'plural',
  operand: { type: 'variable', name: 'count' },
  options: [
    { name: 'minimumFractionDigits', value: { type: 'nmtoken', value: '2' } }
  ]
}
```
For the sake of simplicity, let's say that `$count` may resolve to an externally provided integer value `42`. With that, the resolution could in practice be done by calling the handler for `:plural` that's defined in our registry:
```js
let rv = pluralHandler(42, { minimumFractionDigits: '2'})
```
The important bit here is that `rv` doesn't need to be a concrete formatted value like `'42.00'`, but it could be something like:
```ts
function matchPlurals(
  pr: Intl.PluralRules,
  value: number,
  keys: Set<string>
): string[] {
  const matches: string[] = [];
  const str = String(value);
  if (keys.has(str)) matches.push(str);
  const cat = pr.select(value);
  if (keys.has(cat)) matches.push(cat);

  return matches;
}

// The `bind()` returns a curried function with a single argument `keys`
let rv = matchPlurals.bind(null,
  new Intl.PluralRules('en', { minimumFractionDigits: 2 }),
  42
);
```

So while this does do a little bit of early binding for resolving `$count` and `:plural`, as well as constructing a PluralRules instance, it still leaves the actual selection to be performed later. This is ultimately necessary to ensure that we emit the right errors at the right time. 

For example, if the value given as the `minimumFractionDigits` option were instead `-1`, the `new Intl.PluralRules()` call would have failed with a range error. Our selection method should ensure that an error for this is emitted exactly once, and at an appropriate time compared to errors in any other selector.

It's important to note that this doesn't lead to the requirement that an implementation must be as eager in its resolution, as long as its observable behaviour (including error emission) matches the method presented here.